static/app/utils/number/ladder.tsx

```diff
@@ -0,0 +1,63 @@
+import orderBy from 'lodash/orderBy';
+
+type Rung<T> = [threshold: number, value: T];
+
+/**
+ * Class that represents a value ladder. Given a ladder of increasing thresholds,
+ * the ladder can match an incoming value against the known intervals.
+ *
+ * @example
+ * ```javascript
+ * // Selecting a quota based on team size
+ * const ladder = new Ladder([
+ *   [0, 10],
+ *   [10, 100_000],
+ *   [30, 1_000_000],
+ * ]);
+ *
+ * const quota = ladder.step(0); // Quota for small teams is 10
+ * const quota = ladder.step(12); // Quota for a 10-20 person team is 100,000
+ * const quota = ladder.step(120); // Quota for a 30 person or higher team is 1,000,000
+ * ```
+ */
+export class Ladder<T> {
+  rungs: Rung<T>[];
+
+  // At least one rung is required by the type
+  constructor(rungs: [Rung<T>, ...Rung<T>[]]) {
+    if (!rungs.some(rung => rung[0] === 0)) {
+      throw new Error('At least one rung in the ladder must start at 0');
+    }
+
+    const sortedRungs = orderBy(rungs, rung => rung[0], 'desc');
+
+    const rungThresholds = sortedRungs.map(rung => rung[0]);
+    const uniqueThresholds = new Set(rungThresholds);
+
+    if (uniqueThresholds.size !== rungThresholds.length) {
+      throw new Error('Rung thresholds are not unique');
+    }
```

Comment:
There might be cases where people use dynamically constructed values they pass to this ladder class. In this case, it will be painful for them to debug `Rung threshold are not unique` as it will require them to run the code to see what rungs might not be unique. How about changing this to error with `Attempted to initialize ladder with duplicate rung(s) x,y,z`

Reply:
Roger, added more detailed error checking